### PR TITLE
Fixes to patient-diagnosis-biosource-biomaterial exporter

### PR DESCRIPTION
**Pull request by forus**
TMT-531 Solve issue where studyID is modified by the export/packer
TMT-547 Fix potential bug with value propagation in the export table

### CODE REVIEW

File: packer/table_transformations/patient_diagnosis_biosource_biomaterial.py [hunk 1]
```diff
@@ -47,77 +48,66 @@ def from_obs_df_to_pdbb_df(obs):
...
+            else:
+                break
+        if not row_copied:
+            result_rows.append(row)
+    return pd.DataFrame(result_rows)
+
+
+def _is_ancestor_row(ancestor_row_candidate, descendant_row_candidate, id_columns):
+    for id_column in id_columns:
+        if pd.isna(ancestor_row_candidate[id_column]):
```
Comment (ewelinagr) [inline packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: I have an error here, did you mean:
```suggestion
        if np.isnan(ancestor_row_candidate[id_column]):
```
?

File: packer/table_transformations/patient_diagnosis_biosource_biomaterial.py [hunk 2]
```diff
@@ -47,77 +48,66 @@ def from_obs_df_to_pdbb_df(obs):
...
 
 
 def update_datatypes(data):
     for col in data.columns:
-        # update integer fields - downcasted to float by ffill function
-        # (NaN does not have an integer representation)
-        data[col] = data[col].apply(to_int)
         # update datetime fields
-        if re.match(r'^[0-9]+\.\sDate', col, flags=re.IGNORECASE):
+        if re.match(r'.*[^\\]*\bdate\b[^\\]*\\$', col, flags=re.IGNORECASE):
```
Comment (ewelinagr) [inline packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: Why do you want to change this regex? the new one does not seem to match e.g. `01. date of biomaterial`

Comment (ewelinagr) [reply packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: https://regex101.com/

File: packer/table_transformations/patient_diagnosis_biosource_biomaterial.py [hunk 3]
```diff
@@ -47,77 +48,66 @@ def from_obs_df_to_pdbb_df(obs):
...
             data[col] = data[col].apply(to_datetime)
+        elif np.issubdtype(data[col].dtype, np.number):
```
Comment (ewelinagr) [inline packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: Are you sure it is working in all the cases? For example if there are 'Nan's in a column, the whole column probably does not have the numeric type. The previous way was checking the same, but per cell, not per column

Comment (forus) [reply packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: I think 'Nan's are fine. They are valid values of the `float` type.
`test_values_propagation` test data has `...\\Number\\` concepts with `None` that end up as `NAN` in the column. The test shows that `5.` becomes `5` at the end.

File: packer/table_transformations/patient_diagnosis_biosource_biomaterial.py (hunk 1 above)
Comment (forus) [reply packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: hm, `pd.na` sounds to me like it does broader checks including `None`.
What kind of error?

Comment (ewelinagr) [reply packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: > AttributeError: module 'pandas' has no attribute 'isna'

Comment (ewelinagr) [reply packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: https://stackoverflow.com/a/48313152

Comment (forus) [reply packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: ah, thanks!

Comment (forus) [reply packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: Addressed it here https://github.com/thehyve/transmart-packer/pull/3/commits/6f86c4c2dfd95d351e118f043f5d37446f5f4a65

File: packer/table_transformations/patient_diagnosis_biosource_biomaterial.py [hunk 4]
```diff
@@ -37,87 +36,75 @@ def from_obs_df_to_pdbb_df(obs):
...
-    data.sort_values(id_columns, na_position='first', inplace=True)
-    grouped_data = data.groupby(id_columns[0])
-    # propagate data to lower levels
-    ffill_data = grouped_data.ffill()
-    # limit rows to the lowest level
-    limit_rows_to_lowest_level(ffill_data, id_columns)
-    # drop rows with duplicated identifying columns, keep only the last one
-    ffill_data.drop_duplicates(subset=id_columns, keep='last', inplace=True)
-    return ffill_data
+def _copy_missing_value_to_descendant_row(ancestor_row, descendant_row, id_columns):
```
[the remark this answers is not in the document]
Comment (forus) [reply packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: ok, I'll add underscore prefix to the rest of the internal functions then. Just to communicate that they were not designed to be reused from the outside.

[the remark this answers is not in the document]
Comment (forus) [reply packer/table_transformations/patient_diagnosis_biosource_biomaterial.py]: It's addressed here https://github.com/thehyve/transmart-packer/pull/3/commits/b243a2f59c4ccc3f28c42d6be63c3454b0484d5b

**Pull request merged**